### ISSUE TITLE
Sensitive data masking utility

### ISSUE THREAD

**Issue by heitorlessa** [awslabs/aws-lambda-powertools-roadmap#12]
**Runtime e.g. Python, Java, all of them.** Python

**Is your feature request related to a problem? Please describe.**

As a customer, I'd like to obfuscate incoming data for known fields that contain PII, so that they're not passed downstream or accidentally logged.

**Describe the solution you'd like**

It could be any of these ideas or something better if anyone wants to chime in:

* Similar to [DAZN Obsfucater Middleware for NodeJS](https://github.com/getndazn/dazn-lambda-powertools/tree/master/packages/lambda-powertools-middleware-obfuscater)
* Standalone function that takes a JSON Path and does an in-place editing to mask data using masking of my choice (e.g. `***`, `###`)

A more complex operation would be for data depersonalization where I'd want to encrypt and store the correct data somewhere, in which a separate actor would have permission to decrypt it.

**Describe alternatives you've considered**

1. Mask data on my own with a function that walks through a graph and recursively change it.
2. Use Amazon Comprehend Medical to detect PHI data, convert it to a `str`, and run a string replacement to mask any sensitive data encountered

**Is this something you'd like to contribute if you had guidance?**

**Additional context**

**Comment by michaelbrewer** [comment]
@heitorlessa - would you be able to help put more details on how this could be implemented? It might be something i can have a look into.

**Comment by heitorlessa** [comment]
I couldn't block time to speak to a few Security experts on the best way to do this. One thing I know for sure is that not only you want to mask data, but there are times you want to encrypt it from one party and decrypt only from another party -- it could be optional.

Recursion works, AST also works, JMESPath wouldn't work as you can't do an in-place edit nor know the nodes you travelled.

Let me ping some Security experts to get their opinion on it.

**Comment by heitorlessa** [comment]
> My idea was something like this:
> 
> ```python
> from aws_lambda_powertools import Logger
> from aws_lambda_powertools import obfuscation_filter
> 
> logger = Logger(service="payment")
> 
> @logger.inject_lambda_context
> def lambda_handler(event, context):
> 
>     logger.info(event)
>     #   {
>     #     Records: [
>     #         { firstName: "personal" secondName: "identifiable" email: "inform@ti.on", groupID: "123" },
>     #         { firstName: "second" secondName: "personal" email: "inform@ti.on", some_other_value: "abc" }
>     #       ]
>     #   }
> 
>     # pass object and list of json paths to obfuscate
>     obfuscation_filter(
>         object=event,
>         obfuscation_paths=["Records.*.firstName", "Records.*.secondName", "Records.*.email"]
>     )
> 
>     logger.info(event)
>     # {
>     #   Records: [
>     #       { firstName: "********" secondName: "************" email: "******@**.**", groupID: "123" },
>     #       { firstName: "******" secondName: "********" email: "******@**.**", some_other_value: "abc" }
>     #     ]
>     # }
> ```
> 
> thoughts?

That's a similar UX we would be looking at. I've pinged two security experts I trust at AWS to help brainstorm this, but here's essentially the problem space we're looking at.

1. As a customer, I'd like to mask sensitive data my function is receiving so that downstream systems will not need any additional work on handling PII data.
2. As a customer, I'd like to recover sensitive data masked into its original form so that I can handle sensitive requests around that data on a as-needed basis.


We can handle 1 without any tokenization process, however if we ever want to support 2 we would need tokenization to obfuscate and de-obfuscate data in separate operations, and the latter being done only by authorised personnel.

There could be simpler ways to do it without bringing a heavy dependency but given tokenisation is serious thing and must be uniquely random we need to be careful, hence some Security experts opinion first.... or else we might end up with 1 only that's similar to what you did @keithrozario

**Comment by keithrozario** [comment]
True, my understanding of tokenization is that it requires a store of data (unlike encryption). It'll be quite difficult to execute either tokenization or encryption without some external service providing that functionality.

**Comment by michaelbrewer** [comment]
I might look a little more into this. Especially once we feel like the UX it there for this.

**Comment by faboulaye** [comment]
Any update about this issue ?

**Comment by dreamorosi** [comment]
> Any update about this issue ?

Hi @faboulaye, please see [#1076](https://github.com/awslabs/aws-lambda-powertools-python/issues/1076) for info on the current status of AWS Lambda Powertools for Python. 

For the TypeScript version we have a similar feature request ([#728](https://github.com/awslabs/aws-lambda-powertools-typescript/issues/728)) but at the moment we are focused on refining the experience for the core utilities (Logger, Tracer, Metrics) before considering new feature requests.
